lightning/src/offers/flow.rs

```diff
@@ -0,0 +1,1168 @@
+		&self, invoice_request: InvoiceRequest, context: Option<OffersContext>,
+	) -> Result<VerifiedInvoiceRequest, ()> {
+		let secp_ctx = &self.secp_ctx;
+		let expanded_key = &self.inbound_payment_key;
+
+		let nonce = match context {
+			None if invoice_request.metadata().is_some() => None,
+			Some(OffersContext::InvoiceRequest { nonce }) => Some(nonce),
+			_ => return Err(()),
+		};
+
+		let invoice_request = match nonce {
+			Some(nonce) => {
+				match invoice_request.verify_using_recipient_data(nonce, expanded_key, secp_ctx) {
+					Ok(invoice_request) => invoice_request,
+					Err(()) => return Err(()),
+				}
+			},
+			None => match invoice_request.verify_using_metadata(expanded_key, secp_ctx) {
+				Ok(invoice_request) => invoice_request,
+				Err(()) => return Err(()),
+			},
+		};
+
+		Ok(invoice_request)
+	}
+
+	/// Verifies a [`Bolt12Invoice`] using the provided [`OffersContext`] or the invoice's own metadata,
+	/// returning the corresponding [`PaymentId`] if successful.
+	///
+	/// - If an [`OffersContext::OutboundPayment`] with a `nonce` is provided, verification is performed
+	///   using the payer's context data.
+	/// - If no context is provided and the invoice corresponds to a [`Refund`] without blinded paths,
+	///   verification is performed using the invoice's metadata.
+	/// - If neither condition is met, verification fails.
+	pub fn verify_bolt12_invoice(
+		&self, invoice: &Bolt12Invoice, context: Option<&OffersContext>,
+	) -> Result<PaymentId, ()> {
+		let secp_ctx = &self.secp_ctx;
+		let expanded_key = &self.inbound_payment_key;
+
+		match context {
+			None if invoice.is_for_refund_without_paths() => {
+				invoice.verify_using_metadata(expanded_key, secp_ctx)
+			},
+			Some(&OffersContext::OutboundPayment { payment_id, nonce, .. }) => {
+				invoice.verify_using_payer_data(payment_id, nonce, expanded_key, secp_ctx)
+			},
+			_ => Err(()),
+		}
+	}
+
+	/// Verifies the provided [`AsyncPaymentsContext`] for an inbound [`HeldHtlcAvailable`] message.
+	///
+	/// The context is verified using the `nonce` and `hmac` values, and ensures that the context
+	/// has not expired based on `path_absolute_expiry`.
+	///
+	/// # Errors
+	///
+	/// Returns `Err(())` if:
+	/// - The HMAC verification fails for inbound context.
+	/// - The inbound payment context has expired.
+	///
+	/// [`HeldHtlcAvailable`]: crate::onion_message::async_payments::HeldHtlcAvailable
+	#[cfg(async_payments)]
+	pub fn verify_inbound_async_payment_context(
+		&self, context: AsyncPaymentsContext,
+	) -> Result<(), ()> {
+		match context {
+			AsyncPaymentsContext::InboundPayment { nonce, hmac, path_absolute_expiry } => {
+				signer::verify_held_htlc_available_context(nonce, hmac, &self.inbound_payment_key)?;
+
+				if self.duration_since_epoch() > path_absolute_expiry {
+					return Err(());
+				}
+				Ok(())
+			},
+			_ => Err(()),
+		}
+	}
+
+	/// Verifies the provided [`AsyncPaymentsContext`] for an outbound [`HeldHtlcAvailable`] message.
```

Comment:
s/outbound HeldHtlcAvailable/inbound ReleaseHeldHtlc

Reply:
Thanks, Val, for the pointer!

I got a bit confused about the distinction between `outbound HeldHtlcAvailable` and `inbound ReleaseHeldHtlc`, so I’ve updated the documentation based on my current understanding.

It would be awesome if you could take a look at the [diff](https://github.com/shaavan/rust-lightning/compare/pr3639.15..pr3639.16) and let me know if I’ve got it right.
Thanks again! :)